SimMuon/GEMDigitizer/src/ME0ReDigiProducer.cc

```diff
@@ -13,265 +13,406 @@
-            newTime = iBunch * 25;
-            break;
-          }
-        }
-      }
-
-      edm::LogVerbatim("ME0ReDigiProducer")
-        << "\tBX " << newTime << std::endl;
-
-      // calculate the position in global coordinates
-      const LocalPoint oldLP(me0Digi.x(), me0Digi.y(), 0);
-      const GlobalPoint oldGP(roll->toGlobal(oldLP));
-      const GlobalPoint centralGP(roll->toGlobal(LocalPoint(0.,0.,0.)));
-      const std::vector<float> parameters(roll->specs()->parameters());
-      const float height(parameters[2]); // G4 uses half-dimensions!
-
-      // smear the new radial with gaussian
-      const float oldR(oldGP.perp());
-
-      float newR = oldR;
-      if (me0Digi.prompt() and smearRadial_  and nPartitions_  > 1)
-	newR = CLHEP::RandGaussQ::shoot(engine, oldR, radialResolution_);
-      
-      // calculate the new position in local coordinates
-      const GlobalPoint radialSmearedGP(GlobalPoint::Cylindrical(newR, oldGP.phi(), oldGP.z()));
-      LocalPoint radialSmearedLP = roll->toLocal(radialSmearedGP);
-      
-      // new y position after smearing
-      const float targetYResolution(sqrt(newYResolution_*newYResolution_ - oldYResolution_ * oldYResolution_));
-      float newLPy = radialSmearedLP.y();
-      if (me0Digi.prompt())
-	newLPy = CLHEP::RandGaussQ::shoot(engine, radialSmearedLP.y(), targetYResolution);
-      
-      const ME0EtaPartition* newPart = roll;
-      LocalPoint newLP(radialSmearedLP.x(), newLPy, radialSmearedLP.z());
-      GlobalPoint newGP(newPart->toGlobal(newLP));
-      	
-      // check if digi moves one up or down roll
-      int newRoll = detId.roll();
-      if (newLP.y() > height)  --newRoll;
-      if (newLP.y() < -height) ++newRoll;
-
-      if (newRoll != detId.roll()){
-	// check if new roll is possible
-	if (newRoll < ME0DetId::minRollId || newRoll > ME0DetId::maxRollId){
-	  newRoll = detId.roll();
+
+    /*
+      Starting form the incoming pseudo-digi, which has perfect time and position resolution:
+      1A. Smear time using sigma_t by some value
+      1B. Correct the smeared time with the central arrival time for partition
+      1C. Apply discretization: if the smeared time is outside the BX window (-12.5ns;+12.5ns),
+      the hit should be assigned to the next (or previous) BX
+
+      2A. Find strip that the digi belongs to
+      2B. Get the center of this strip and the error on the position assuming the geometry
+
+      3A. Filter event if a digi at this partition/strip/BX already exists
+      3B. Add to collection
+    */
+
+	LogDebug("ME0ReDigiProducer::buildDigis") << "Begin building digis."<<std::endl;
+	ME0DigiPreRecoCollection::DigiRangeIterator me0dgIt;
+	for (me0dgIt = input_digis.begin(); me0dgIt != input_digis.end();
+			++me0dgIt){
+
+		const auto& me0Id = (*me0dgIt).first;
+		LogTrace("ME0ReDigiProducer::buildDigis") << "Starting with roll: "<< me0Id<<std::endl;
+
+		//setup map for this chamber/eta partition
+		ChamberDigiMap chDigiMap;
```

Comment:
@davidlange6 I changed it to a map of a tuple. I think it that the map is better than a vector in this case because when we create each digi we have to check if it has already been created.